File: leetcode/0015.3Sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        copy = {}
        res = []
        for i in range(len(nums) - 1):
            start = i+1
            end = len(nums) - 1
            while start < end:
                total = nums[i] + nums[start] + nums[end]
                if total < 0:
                    start += 1
                elif total > 0:
                    end -= 1
                else:
                    var = str(nums[i]) + str(nums[start]) + str(nums[end])
                    if var not in copy:
                        tmpArr = [nums[i], nums[start], nums[end]]
                        res.append(tmpArr)
                    copy[var] = 1
                    start += 1
                    while( nums[start] == nums[start-1] and start < end):
                        start += 1
                    #if var not in copy:
        return (res)

File: leetcode/0015.3-sum.ts
function threeSum(nums: number[]): number[][] {
    const result = []
    const map = {}
    nums.sort((a, b) => a - b); 

    for(let i=0; i<nums.length; i++){
        let l = i+1;
        let r = nums.length -1

        while(l<r) {
            let temp = nums[i] + nums[l] + nums[r]
            if(temp < 0) {
                l += 1
            }
            else if(temp > 0) {
                r -= 1
            }
            else {
                let res = [nums[i], nums[l], nums[r]]
                const mapKey = `${nums[i]},${nums[l]},${nums[r]}`; 
                if(map[mapKey] === undefined) {
                    map[mapKey] = 1
                    result.push(res)
                }

                l += 1
            }

        }
    }
    return result
};